File: samples/python/agents/signing_and_verifying/test_client.py
import asyncio
import logging

import httpx

from a2a.client import A2ACardResolver, ClientConfig, create_client
from a2a.helpers import display_agent_card
from a2a.types import GetExtendedAgentCardRequest
from a2a.utils.constants import (
    AGENT_CARD_WELL_KNOWN_PATH,
)
from a2a.utils.signing import create_signature_verifier
from cryptography.hazmat.primitives import serialization
from jwt.api_jwk import PyJWK


def _key_provider(kid: str, jku: str) -> PyJWK | str | bytes:
    """Fetch and parse public key from JKU URL given key ID (kid) and JKU URL."""
    if not isinstance(kid, str) or not kid:
        raise TypeError(f'Expected kid: str, but got: {type(kid).__name__} ({kid!r})')
    if not isinstance(jku, str) or not jku:
        raise TypeError(f'Expected jku: str, but got: {type(jku).__name__} ({jku!r})')

    try:
        response = httpx.get(jku)
        response.raise_for_status()
    except httpx.HTTPError as err:
        raise ValueError(f'Failed to fetch public key from JKU URL ({jku}): {err}') from err

    keys = response.json()
    pem_data_str = keys.get(kid)

    if not pem_data_str:
        raise ValueError('Invalid JWK Key ID.')

    return serialization.load_pem_public_key(pem_data_str.encode('utf-8'))


# Create a verifier function to validate AgentCard JWS signatures
verify_card_signature = create_signature_verifier(_key_provider, ['ES256'])


async def main() -> None:
    """Main function."""

    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)

    base_url = 'http://localhost:9999'

    async with httpx.AsyncClient() as httpx_client:
        # Initialize A2ACardResolver
        resolver = A2ACardResolver(
            httpx_client=httpx_client,
            base_url=base_url,
        )

        try:
            logger.info(
                'Attempting to fetch public agent card from: %s%s',
                base_url,
                AGENT_CARD_WELL_KNOWN_PATH,
            )
            # Pass verify_card_signature to validate the signature on the public Agent Card
            public_card = await resolver.get_agent_card(
                signature_verifier=verify_card_signature,
            )
            logger.info('Successfully fetched public agent card:')

        except Exception as e:
            logger.exception(
                'Critical error fetching public agent card.',
            )
            raise RuntimeError from e

        # Create Base Client directly via unified factory
        client = await create_client(
            agent=public_card,
            client_config=ClientConfig(streaming=False),
        )

        # Pass verify_card_signature to validate the signature on the extended Agent Card
        extended_card_with_signature = await client.get_extended_agent_card(
            GetExtendedAgentCardRequest(),
            signature_verifier=verify_card_signature,
        )
        logger.info('Successfully fetched extended agent card with signature:')
        display_agent_card(extended_card_with_signature)
        logger.info('Signature:')
        logger.info(extended_card_with_signature.signatures)

        extended_card_without_signature = await client.get_extended_agent_card(
            GetExtendedAgentCardRequest()
        )
        logger.info(
            'Successfully fetched extended agent card without signature:'
        )  # Signature is only for client-side verification purpose
        display_agent_card(extended_card_without_signature)


if __name__ == '__main__':
    asyncio.run(main())

File: samples/js/agents/signing_and_verifying/test_client.ts
import * as crypto from 'node:crypto';

import {
  AGENT_CARD_PATH,
  AgentCard,
  verifyAgentCardSignature,
} from '@a2a-js/sdk';
import { ClientFactory, DefaultAgentCardResolver } from '@a2a-js/sdk/client';

async function keyProvider(kid: string, jku?: string): Promise<crypto.KeyObject> {
  /** Fetch and parse public key from JKU URL given key ID (kid) and JKU URL. */
  if (typeof kid !== 'string' || !kid) {
    throw new TypeError(`Expected kid: string, but got: ${typeof kid} (${kid})`);
  }
  if (typeof jku !== 'string' || !jku) {
    throw new TypeError(`Expected jku: string, but got: ${typeof jku} (${jku})`);
  }

  let response: Response;
  try {
    response = await fetch(jku);
    if (!response.ok) {
      throw new Error(`HTTP ${response.status} ${response.statusText}`);
    }
  } catch (err) {
    throw new Error(`Failed to fetch public key from JKU URL (${jku}): ${err}`);
  }

  const keys = (await response.json()) as Record<string, string>;
  const pemDataStr = keys[kid];

  if (!pemDataStr) {
    throw new Error('Invalid JWK Key ID.');
  }

  return crypto.createPublicKey(pemDataStr);
}

// Create a verifier function to validate AgentCard JWS signatures
const verifyCardSignature = verifyAgentCardSignature(keyProvider);

async function main(): Promise<void> {
  /** Main function. */

  const baseUrl = 'http://localhost:9999';

  // Initialize AgentCardResolver
  const resolver = new DefaultAgentCardResolver();

  let publicCard: AgentCard;
  try {
    console.log(
      `Attempting to fetch public agent card from: ${baseUrl}${AGENT_CARD_PATH}`
    );
    publicCard = await resolver.resolve(baseUrl);

    // Pass verifyCardSignature to validate the signature on the public Agent Card
    await verifyCardSignature(publicCard);
    console.log('Successfully fetched public agent card:');
  } catch (e) {
    console.error('Critical error fetching public agent card:', e);
    throw e;
  }

  // Create Client directly via ClientFactory
  const clientFactory = new ClientFactory();
  const client = await clientFactory.createFromAgentCard(publicCard);

  // Pass verifyCardSignature to validate the signature on the extended Agent Card
  const extendedCardWithSignature = await client.getAgentCard(
    undefined,
    verifyCardSignature
  );

  console.log('Successfully fetched extended agent card with signature:');
  console.log(JSON.stringify(extendedCardWithSignature, null, 2));
  console.log('Signature:');
  console.log(JSON.stringify(extendedCardWithSignature.signatures, null, 2));

  const extendedCardWithoutSignature = await client.getAgentCard();
  console.log('Successfully fetched extended agent card without signature:');
  console.log(JSON.stringify(extendedCardWithoutSignature, null, 2));
}

main().catch((err) => {
  console.error('Error in main:', err);
  process.exit(1);
});
